theano/gof/vm.py

```diff
@@ -92,46 +93,56 @@ def calculate_reallocate_info(order, fgraph, storage_map, compute_map_re, depend
                 viewed_by[origin].append(out)
             idx_o += 1
 
-        for ins in node.inputs:
-            assert not (ins in view_of and viewed_by[ins])
-            if (getattr(ins, 'ndim', None) == 0 and not storage_map[ins][0]
-                    and ins not in fgraph.outputs and ins.owner
-                    and all([compute_map_re[v][0] for v in dependencies.get(ins, [])])
-                    and ins not in allocated):
-                # Constant Memory cannot be changed
-                # Constant and shared variables' storage_map value is not empty
-                reuse_out = None
-                if ins not in view_of and not viewed_by.get(ins, []):
-                    # where gc
-                    for i in range(idx + 1, len(order)):
-                        if reuse_out:
-                            break
-                        for out in order[i].outputs:
-                            if (getattr(out, 'ndim', None) == 0 and out not in pre_allocated
-                                    and ins.type == out.type):
-                                reuse_out = out
-                                pre_allocated.add(out)
-                                allocated.add(ins)
-                elif ins in view_of:
-                    origin = view_of[ins]
-                    if ins in viewed_by[origin]:
-                        viewed_by[origin].remove(ins)
-                    if (not viewed_by[origin] and
-                            origin not in fgraph.inputs and
-                            not isinstance(origin, theano.Constant)):
+        if hasattr(node.fgraph, 'shape_feature'):
```

Comment:
This make the inner code more indented (so harder to read sometimes due to max line length of 80 characters) and disable it for ndim == 0.

I think doing it like this is better:
```
if hasattr(node.fgraph, 'shape_feature'):
    same_shape = node.fgraph.shape_feature.same_shape
else:
    def same_shape(x, y):
        if x.ndim == 0 and y.ndim == 0:  return True
        else: return False
```

Then you can keep the old indentation and it still work for ndim == 0.